### PR TITLE
use atomic file copy when writing to posix file system

### LINKED ISSUE #2897: clarify whether JobStore.updateFile is atomic or not
Code in Cactus currently assumes that JobStore.updateFile is atomic, but this appears not to always be the case in practice, since the fileJobStore implementation just uses Python's `shutil.copyfile` (with no further interlocks).

It would be useful to clarify the intended behavior.

Here's the related Cactus code: https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/pipeline/cactus_workflow.py#L301-L309

See also https://github.com/ComparativeGenomicsToolkit/cactus/issues/114

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-474)
┆Issue Number: TOIL-474

### PR DESCRIPTION
**Pull request by diekhans**
Use atomic file copy when writing to posix file system.  
fixes #2897

This is ready to go now,

### CODE REVIEW

File: src/toil/lib/misc.py [hunk 1]
```diff
@@ -191,3 +192,62 @@ def truncExpBackoff():
...
+    "atomic install of tmp_path as final_path"
+    if os.path.dirname(os.path.normpath(final_path)) != '/dev':
+        os.rename(tmp_path, final_path)
+
+
+class AtomicFileCreate(object):
+    """Context manager to great an temporary file.  Entering returns path to the
+    temporary file in the same directory.  If the code in context succeeds,
+    the file renamed to it's actually nape.  If an error occurs, the file is
+    not installed and is removed unless keeps is specified.
```
Comment (michaelkarlcoleman) [inline src/toil/lib/misc.py]: Minor typos:
- s/great/create/
- s/it's/its/
- s/nape/name/
- s/keeps/keep/

Beyond this, this patch looks solid to me!

File: src/toil/lib/misc.py [hunk 2]
```diff
@@ -191,3 +192,62 @@ def truncExpBackoff():
         t *= 2
     while True:
         yield t
+
+
+def atomic_tmp_file(final_path):
+    """Return a tmp file name to use with atomic_install.  This will be in the
+    same directory as final_path. The temporary file will have the same extension
+    as finalPath.  In final path is in /dev (/dev/null, /dev/stdout), it is
```
Comment (mr-c) [inline src/toil/lib/misc.py]:
```suggestion
    as finalPath.  If the final path is in /dev (/dev/null, /dev/stdout), it is
```

File: src/toil/lib/misc.py [hunk 3]
```diff
@@ -191,3 +192,62 @@ def truncExpBackoff():
...
+            try:
+                os.unlink(self.tmp_path)
+            except:
+                pass
+        return False
+
+def atomic_copy(src_path, dest_path):
+    """Copy a file using posix atomic creations semantics."""
+    with AtomicFileCreate(dest_path) as dest_path_tmp:
+        shutil.copyfile(src_path, dest_path_tmp)
```
Comment (mr-c) [inline src/toil/lib/misc.py]: The word `atomic` does not appear in the documentation for `shutil.copyfile`, is there another place that this guarantee is made?

https://docs.python.org/3/library/shutil.html

**Comment by mr-c** [comment]
@diekhans FYI, there are 4+ test failures at https://ucsc-ci.com/databiosphere/toil/-/jobs/33077/raw

**Comment by adamnovak** [comment]
I've pushed this PR's commit onto issues/2897-atomic-jobstore-diekhans for testing.

**Comment by adamnovak** [comment]
Looks like the new function isn't getting imported in one particular place:

```
Traceback (most recent call last):
  File "/builds/databiosphere/toil/src/toil/test/__init__.py", line 626, in fx
    return generalMethod(self, **prms)
  File "/builds/databiosphere/toil/src/toil/test/jobStores/jobStoreTest.py", line 703, in testImportExportFile
    jobStoreFileID = self.jobstore_initialized.importFile(srcUrl)
  File "/builds/databiosphere/toil/src/toil/jobStores/abstractJobStore.py", line 304, in importFile
    return self._importFile(otherCls, srcUrl, sharedFileName=sharedFileName, hardlink=hardlink)
  File "/builds/databiosphere/toil/src/toil/jobStores/aws/jobStore.py", line 426, in _importFile
    sharedFileName=sharedFileName)
  File "/builds/databiosphere/toil/src/toil/jobStores/abstractJobStore.py", line 325, in _importFile
    otherCls._readFromUrl(url, writable)
  File "/builds/databiosphere/toil/src/toil/jobStores/fileJobStore.py", line 314, in _readFromUrl
    atomic_copyobj(readable, writable, length=cls.BUFFER_SIZE)
NameError: name 'atomic_copyobj' is not defined
```

**Comment by diekhans** [comment]
Thanks!  I will update the doc before final merge.

**Comment by diekhans** [comment]
I am close this PR and opening one on the issue branch.

**Pull request closed without merging**